samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,852 @@
+#else
+#error SW0_GPIO_PIN needs to be set
+#endif
+
+#ifdef SW0_GPIO_FLAGS
+#define PIN0_FLAGS SW0_GPIO_FLAGS
+#else
+#error SW0_GPIO_FLAGS needs to be set
+#endif
+
+#ifdef SW1_GPIO_PIN
+#define PIN1	SW1_GPIO_PIN
+#endif
+
+#ifdef SW1_GPIO_CONTROLLER
+#define PORT1	SW1_GPIO_CONTROLLER
+#endif
+
+#ifdef SW1_GPIO_FLAGS
+#define PIN1_FLAGS SW1_GPIO_FLAGS
+#endif
+
+#ifdef SW2_GPIO_PIN
+#define PIN2	SW2_GPIO_PIN
+#endif
+
+#ifdef SW2_GPIO_CONTROLLER
+#define PORT2	SW2_GPIO_CONTROLLER
+#endif
+
+#ifdef SW2_GPIO_FLAGS
+#define PIN2_FLAGS SW2_GPIO_FLAGS
+#endif
+
+#ifdef SW3_GPIO_PIN
+#define PIN3	SW3_GPIO_PIN
+#endif
+
+#ifdef SW3_GPIO_CONTROLLER
+#define PORT3	SW3_GPIO_CONTROLLER
+#endif
+
+#ifdef SW3_GPIO_FLAGS
+#define PIN3_FLAGS SW3_GPIO_FLAGS
+#endif
+
+/* Event FIFO */
+
+K_FIFO_DEFINE(evt_fifo);
+
+enum evt_t {
+	GPIO_BUTTON_0	= 0x00,
+	GPIO_BUTTON_1	= 0x01,
+	GPIO_BUTTON_2	= 0x02,
+	GPIO_BUTTON_3	= 0x03,
+	CDC_UP		= 0x04,
+	CDC_DOWN	= 0x05,
+	CDC_LEFT	= 0x06,
+	CDC_RIGHT	= 0x07,
+	CDC_UNKNOWN	= 0x08,
+	CDC_STRING	= 0x09,
+	HID_MOUSE_CLEAR	= 0x0A,
+	HID_KBD_CLEAR	= 0x0B,
+	HID_KBD_STRING	= 0x0C,
+};
+
+struct app_evt_t {
+	sys_snode_t node;
+	struct k_mem_block block;
+	enum evt_t event_type;
+};
+
+#define FIFO_ELEM_MIN_SZ        sizeof(struct app_evt_t)
+#define FIFO_ELEM_MAX_SZ        sizeof(struct app_evt_t)
+#define FIFO_ELEM_COUNT         255
+#define FIFO_ELEM_ALIGN         sizeof(unsigned int)
+
+K_MEM_POOL_DEFINE(event_elem_pool, FIFO_ELEM_MIN_SZ, FIFO_ELEM_MAX_SZ,
+		  FIFO_ELEM_COUNT, FIFO_ELEM_ALIGN);
+
+static inline void app_evt_free(struct app_evt_t *ev)
+{
+	k_mem_pool_free(&ev->block);
+}
+
+static inline void app_evt_put(struct app_evt_t *ev)
+{
+	k_fifo_put(&evt_fifo, ev);
+}
+
+static inline struct app_evt_t *app_evt_get(void)
+{
+	return k_fifo_get(&evt_fifo, K_NO_WAIT);
+}
+
+static inline void app_evt_flush(void)
+{
+	struct app_evt_t *ev;
+
+	do {
+		ev = app_evt_get();
+		if (ev) {
+			app_evt_free(ev);
+		}
+	} while (ev != NULL);
+}
+
+static inline struct app_evt_t *app_evt_alloc(void)
+{
+	int ret;
+	struct app_evt_t *ev;
+	struct k_mem_block block;
+
+	ret = k_mem_pool_alloc(&event_elem_pool, &block,
+			       sizeof(struct app_evt_t),
+			       K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("APP event allocation failed!");
+		app_evt_flush();
+
+		ret = k_mem_pool_alloc(&event_elem_pool, &block,
+				       sizeof(struct app_evt_t),
+				       K_NO_WAIT);
+		if (ret < 0) {
+			LOG_ERR("APP event memory corrupted.");
+			__ASSERT_NO_MSG(0);
+			return NULL;
+		}
+		return NULL;
+	}
+
+	ev = (struct app_evt_t *)block.data;
+	ev->block = block;
+
+	return ev;
+}
+
+/* HID */
+
+static const u8_t hid_mouse_report_desc[] = HID_MOUSE_REPORT_DESC(2);
+static const u8_t hid_kbd_report_desc[] = HID_KEYBOARD_REPORT_DESC();
+
+static K_SEM_DEFINE(evt_sem, 0, 1);	/* starts off "not available" */
+static K_SEM_DEFINE(usb_sem, 1, 1);	/* starts off "available" */
+static struct gpio_callback callback[4];
```

Comment:
If you wrapped these in a structure with an additional field to keep the associated button (`GPIO_BUTTON_1` etc), you could retrieve that in the callback using `CONTAINER_OF` and avoid the copy-paste in the `btn*` functions, keeping only one.